Fix meta registry in export

tugsbayasgalan opened the pull request:
Summary: Title

Test Plan: CI

This fixes some breaking tests in executorch. I think the root cause is when we have aten::matmul which we are not preserving, we register meta implementation from C++ side. It seems like the C++ kernel doesn't work well with mix of FakeTensor and real tensor. This PR sidesteps this problem by always preferring python CIA decomp over C++ Cia decomp

Differential Revision: D63297050

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -189,19 +176,22 @@ def _override_composite_implicit_decomp(cia_ops_to_callable, safe=True):
                 )
+        # See NOTE: Registering old CIA to Meta kernel
+        # It is important that we cache this before we override py_kernels.
+        orig_cia_callable = _get_decomp_for_cia(op_overload)
```
Oh, I guess `_get_decomp_for_cia` already has the logic to:

(1) run the python dispatcher CIA registration if it exists
(2) run the C++ CIA registration if it exists
(3) error if there is no CIA registration

tugsbayasgalan replied:
Yep it does.

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -211,10 +201,18 @@ def _override_composite_implicit_decomp(cia_ops_to_callable, safe=True):
+            # now will call into the custom user defined CIA which can cause a problem.
+            # Since we only run meta impl for shape prop, we should always use the default
+            # behavior. 
+            op_overload.py_impl(torch._C.DispatchKey.Meta)(orig_cia_callable)
```
Can you add some OSS tests? In particular, the one you had that involved:

(1) the model is using Linear()
(2) you export the model with `aten.linear()` not decomposed
(3) dynamic shapes are enabled

tugsbayasgalan replied:
Added!

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -7222,8 +7222,6 @@ def forward(self, x):
             if node.op == "call_function":
                 self.assertTrue(False)
 
-    @testing.expectedFailureTrainingIRToRunDecomp  # T200904004
```
@bdhirsh  it also fixed some other previosly filed errors lol

tugsbayasgalan committed:
Fix meta registration mechanism in export (#136492)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/136492

Title

Test Plan: CI

Differential Revision: D63297050

bdhirsh commented on test/export/test_export.py:
```diff
@@ -7242,6 +7240,41 @@ def forward(self, spec):
+
+        if IS_FBCODE:
+            ep = ep.run_decompositions(
+                {}, _preserve_ops=(torch.ops.aten.linear.default,)
```
what's the plan for unifying the fbcode path?

tugsbayasgalan replied:
https://www.internalfb.com/diff/D62906375 I have a PR here to remove this. I will probably land it this week.

bdhirsh reviewed (approved):
sgtm!
